design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+The remainder of this one pager will propose a high-level approach for incrementally evolving KSQL from a stream-processing system towards a streaming database, releasing the highest ROI capabilities for users first.
+
+## Goals
+
+* Make KSQL fundamentally more useful to application developers by exposing the data already stored within it to conventional `SELECT` queries.
+* Simplify the infrastructural requirements for interactive queries against materialized rows generated by KSQL.
+
+## Scope
+
+* **In**
+  * The high-level functional capabilities comprising interactive queries will be described.
+  * This high-level functionality will be roughly broken down into progressive tiers of completeness.
+  * The intended user experience of interactive queries will be described.
+* **Out**
+  * No specific engineering solutions to any of the capabilities discussed in this one pager will be proposed.
+  * Performance challenges of interactive queries will not be discussed.
+  * No proposals will be made regarding transactionality/consistency semantics around interactive queries.
+
+## Approach
+
+There is a wide spectrum of complexity between minimally useful row lookups and arbitrarily complex SQL queries over materialized state. Adding support for even the most trivial queries will still add significant value for KSQL users, so we should consider interactive queries an ongoing, incremental effort spanning numerous KSQL releases.
+
+Initial limitations around the kinds of interactive queries KSQL can support will also be augmented by the flexibility that users still have when defining materialized views with continuous queries. For example, even if aggregates are not initially supported for interactive queries, they can still be used in a materialized view’s definition. The same is true for joins.
+
+Here we roughly subdivide the spectrum of interactive query complexity into five progressive tiers, each tier building on the one before it. Each of these tiers can be released separately, although it may make sense to combine some of them.
+
+*Please note that some of the example queries below use mock functions for illustrative purposes.*
+
+### Tier 0 - Simple key-based lookups
+
+Support for key-based lookups is the bare minimum starting point for interactive queries that will still add meaningful utility. This essentially amounts to a cosmetic SQL layer above RocksDB’s key-value interface. Key-based lookups will allow users to look up a single row, given a row key:
+
+* WHERE clause must be strictly of the form: `WHERE ROWKEY = <rowkey>`
+* Each query can therefore only target a single shard.
```

Comment:
following on from Vinoth's comment on the use of `shard`... I think a lot of the time we may be talking not about a rocks db shard, (for which each partition/task has multiple), but more about KSQL nodes.  i.e. can a request be serviced by a single node, or several, or all?

I don't think it's that important to be pedantic about this in the KLIP, but thought I'd mention it from @vinothchandar